src/Pos/Generator/BlockEvent.hs

```diff
@@ -41,13 +91,334 @@ data BlockRollbackResult
+                                        over _NewestFirst select
+                                return . BlkEvRollback $ BlockEventRollback
+                                    { _berInput    = blockIndices
+                                    , _berOutValid = BlockRollbackFailure
+                                    }
+            (True, False) -> do
+                -- Fail without rollback (with apply).
+                aft <- getRandom
+                case aft of
+                    AftApplyBad -> do
+                        -- Attempt to apply an invalid sequence of blocks.
+                        let tip = NE.last (getOldestFirst blockchain)
+                        blockIndices <- genBlockIndices (tip + 1)
+                        let
+                            blockIndices' =
+                                -- Those block indices are broken by construction
+                                -- because we append the current tip to the end.
+                                -- Sometimes we can violate the maximum bound on
+                                -- block count, but this is fine. See NOTE
+                                -- on reordering corner cases.
+                                over _OldestFirst (Smg.<> pure tip) blockIndices
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices'
+                            , _beaOutValid = BlockApplyFailure
+                            }
+                    AftApplyNonCont -> do
+                        -- Attempt to apply blocks which are not a valid
+                        -- contuniation of our chain.
+                        let
+                            tip  = NE.last (getOldestFirst blockchain)
+                            -- The amount of blocks to skip. One is enough.
+                            skip = 1
+                        blockIndices <- genBlockIndices (tip + 1 + skip)
+                        return . BlkEvApply $ BlockEventApply
+                            { _beaInput    = blockIndices
+                            , _beaOutValid = BlockApplyFailure
+                            }
+            (False, True) -> do
+                -- Success with rollback.
+                len <- getRandomR (1, length blockchain)
+                let
+                    -- 'NE.fromList' is valid here because 'len >= 1'.
+                    select = NE.fromList . NE.take len
+                    blockIndices = toNewestFirst blockchain &
+                        over _NewestFirst select
+                return . BlkEvRollback $ BlockEventRollback
+                    { _berInput    = blockIndices
+                    , _berOutValid = BlockRollbackSuccess
+                    }
+            (False, False) -> do
+                -- Success without rollback (with apply).
+                let tip = NE.last (getOldestFirst blockchain)
+                blockIndices <- genBlockIndices (tip + 1)
```

Comment:
Also I don't understand how you ensure that total number of blocks can't exceed `begpBlockCountMax`. For instance, here we may already have some blocks. And then `genBlockIndices` can generate `begpBlockCountMax + 1` indices. And all these indices will be in the result.
Will we trim them somewhere?